ndc-models/src/lib.rs

```diff
@@ -420,9 +420,37 @@ pub struct Query {
     pub offset: Option<u32>,
     pub order_by: Option<OrderBy>,
     pub predicate: Option<Expression>,
+    ///
+    pub groups: Option<Grouping>,
 }
 // ANCHOR_END: Query
 
+// ANCHOR: Grouping
+#[skip_serializing_none]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[schemars(title = "Grouping")]
+pub struct Grouping {
+    /// Dimensions along which to partition the data
+    pub dimensions: IndexMap<String, Dimension>,
+    /// Fields to fetch in each group
+    pub fields: IndexMap<String, Field>,
+    /// Aggregates to compute in each group
+    pub aggregates: IndexMap<String, Aggregate>,
+}
```

Comment:
I think we also need to add `having`, `limit`, `offset` and group `order_by` concepts here too.

Reply:
`having` is the only one that's new, right? I have a separate JIRA issue for that.

The others are specified already on the `Query` level, unless you mean sorting and paginating the _groups_ themselves, but I don't think that's a common pattern?

Reply:
Are we taking about ordering and paginating the _groups_ themselves, e.g. order by count-of-group? So then we'd have two order by's, two limits, and two offsets?

Reply:
> Are we taking about ordering and paginating the groups themselves, e.g. order by count-of-group? So then we'd have two order by's, two limits, and two offsets?

Yes.

You can filter, order and paginate the input to your grouping ([see here](https://github.com/hasura/v3-engine/blob/main/rfcs/aggregations.md#filtering-ordering-and-paginating-the-objects-grouped-over)), and then after the grouping and group aggregates are performed over those input rows, you can filter the groups (having), order them, and paginate them.

Here's a fairly nonsensical example, but it uses all the features:

```graphql
  query MyQuery {
    Invoice_groups(
      filter_input: {
        where: {BillingAddress: {Country: {_eq: "Australia"}}}, 
        order_by: {BillingAddress: {State: Asc}}, 
        offset: 1, 
        limit: 10
      }
      grouping_keys: {Customer: {_scalar_field: LastName}}
      having: {_count: {_gt: 2}}
      order_by: {group_key: {Customer: {FirstName: Asc}}}
      limit: 5
      offset: 1
    ) {
      group_key {
        Customer {
          LastName
        }
      }
      group_aggregate {
        _count
      }
    }
  }

```
(play with it [here](https://cloud.hasura.io/public/graphiql?endpoint=https%3A%2F%2Fhasura-v3-aggregates-mock.deno.dev&query=query+MyQuery+%7B%0A++Invoice_groups%28%0A++++filter_input%3A+%7B%0A++++++where%3A+%7BBillingAddress%3A+%7BCountry%3A+%7B_eq%3A+%22Australia%22%7D%7D%7D%2C+%0A++++++order_by%3A+%7BBillingAddress%3A+%7BState%3A+Asc%7D%7D%2C+%0A++++++offset%3A+1%2C+%0A++++++limit%3A+10%0A++++%7D%0A++++grouping_keys%3A+%7BCustomer%3A+%7B_scalar_field%3A+LastName%7D%7D%0A++++having%3A+%7B_count%3A+%7B_gt%3A+2%7D%7D%0A++++order_by%3A+%7Bgroup_key%3A+%7BCustomer%3A+%7BFirstName%3A+Asc%7D%7D%7D%0A++++limit%3A+5%0A++++offset%3A+1%0A++%29+%7B%0A++++group_key+%7B%0A++++++Customer+%7B%0A++++++++LastName%0A++++++%7D%0A++++%7D%0A++++group_aggregate+%7B%0A++++++_count%0A++++%7D%0A++%7D%0A%7D%0A))

Reply:
I've added `HAVING` to this PR, but I haven't finished order by yet.

Reply:
Order-by is here now as well.